pkg/rserverless/rserverless.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2018-2023 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package rserverless
+
+import (
+	"fmt"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+)
+
+// Service represents a serverless service.
+type Service interface {
+	Start()
+}
+
+// Services is a map of service name and its new function.
+var Services = map[string]NewService{}
+
+// Register registers a new serverless service with name and new function.
+func Register(name string, newFunc NewService) {
+	Services[name] = newFunc
+}
+
+// NewService is the function that serverless services need to register at init time.
+type NewService func(conf map[string]interface{}, log *zerolog.Logger) (Service, error)
+
+// Serverless contains the serveless collection of services.
+type Serverless struct {
+	conf     *config
+	log      zerolog.Logger
+	services map[string]Service
+}
+
+type config struct {
+	Services map[string]map[string]interface{} `mapstructure:"services"`
+}
+
+// New returns a new serverless collection of services.
+func New(m interface{}, l zerolog.Logger) (*Serverless, error) {
+	conf := &config{}
+	if err := mapstructure.Decode(m, conf); err != nil {
+		return nil, err
+	}
+
+	n := &Serverless{
+		conf:     conf,
+		log:      l,
+		services: map[string]Service{},
```

Comment:
Why do you need this map?

Reply:
Indeed, this requires clarification.

This is there to follow the other types of services ([`rhttp`](https://github.com/cs3org/reva/blob/master/pkg/rhttp/rhttp.go#L69) and [`rgrpc`](https://github.com/cs3org/reva/blob/master/pkg/rgrpc/rgrpc.go#L122C5)). It is unused right now, but I'll make use of it for the graceful stop.